File: Java/src/main/java/drawing/shapes/Shape.java
package drawing.shapes;

import drawing.writing.JPEGWriter;
import drawing.writing.PNGWriter;

import java.io.IOException;
import java.io.Writer;

public interface Shape {

    Line[] toLines();

    default void draw(Writer writer, Line[] lines) {
        try {
            for (Line line : lines) {
                if (writer instanceof JPEGWriter) {
                    writer.write(line.toJPEG());
                }
                else if (writer instanceof PNGWriter) {
                    writer.write(line.toPNG());
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: TypeScript/src/drawing/shapes/shape.ts
import { Writer } from "../writing/writer";
import { PNGWriter } from "../writing/pngwriter";
import { JPEGWriter } from "../writing/jpegwriter";
import { Line } from "./line";

interface Shape {
	
    toLines(): Line[];

    draw(writer: Writer, lines: Line[]): void;
}

abstract class AbstractShape implements Shape {

	abstract toLines(): Line[];

	public draw(writer: Writer, lines: Line[]): void {
		try {
			for (let line of lines) {
				if (writer instanceof JPEGWriter) {
					writer.write(line.toJPEG());
				}
				else if (writer instanceof PNGWriter) {
					writer.write(line.toPNG());
				}
			}
		} catch (e) {
			console.log(e);
		}
	}
}

export { Shape, AbstractShape }